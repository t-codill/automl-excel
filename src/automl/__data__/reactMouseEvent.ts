// tslint:disable-next-line: no-any
export const reactMouseEvent = {} as unknown as React.MouseEvent<any>;
