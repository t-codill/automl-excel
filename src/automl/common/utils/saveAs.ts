export const saveAs = (content: Blob, fileName: string): void => {
    const evt = document.createEvent("MouseEvents");
    evt.initMouseEvent("click", true, true, window, 0, 0, 0, 0, 0, false, false, false, false, 0, null);

    const link = window.document.createElement("a");
    link.setAttribute("href", window.URL.createObjectURL(content));
    link.setAttribute("download", fileName);
    link.dispatchEvent(evt);
};
