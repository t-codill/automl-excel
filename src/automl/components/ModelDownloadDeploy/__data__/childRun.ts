import { RunHistoryAPIsModels } from "@vienna/runhistory";

export const childRun: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto = {
    runId: "test-run-id_1",
    properties: {
        DataPrepJsonString: "{\"features\":[\"timeStamp\",\"precip\",\"temp\"]}"
    }
};
