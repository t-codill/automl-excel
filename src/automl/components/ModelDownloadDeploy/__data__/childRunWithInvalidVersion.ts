import { RunHistoryAPIsModels } from "@vienna/runhistory";

export const childRunWithInvalidVersion: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto = {
    runId: "test-run-id_1",
    properties: {
        dependencies_versions: "{\"test\": \"1.0.0\"}"
    }
};
