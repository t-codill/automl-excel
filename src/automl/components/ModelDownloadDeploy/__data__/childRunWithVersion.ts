import { RunHistoryAPIsModels } from "@vienna/runhistory";

export const childRunWithVersion: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto = {
    runId: "test-run-id_1",
    properties: {
        dependencies_versions: "{\"azureml-train-automl\": \"1.0.0\"}"
    }
};
