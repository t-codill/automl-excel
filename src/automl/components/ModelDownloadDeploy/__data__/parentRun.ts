import { RunHistoryAPIsModels } from "@vienna/runhistory";

export const parentRun: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto = {
    runId: "test-run-id",
    properties: {
        DataPrepJsonString: "{\"features\":[\"timeStamp\",\"precip\",\"temp\"]}"
    }
};
