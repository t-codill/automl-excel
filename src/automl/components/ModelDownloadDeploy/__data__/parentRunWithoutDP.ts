import { RunHistoryAPIsModels } from "@vienna/runhistory";

export const parentRunWithoutDP: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto = {
    runId: "test-run-id",
    properties: {
    }
};
