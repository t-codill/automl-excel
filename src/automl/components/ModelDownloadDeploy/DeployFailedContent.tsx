import { Link, MessageBar, MessageBarType } from "office-ui-fabric-react";
import * as React from "react";
import { saveAs } from "../../common/utils/saveAs";
import { ModelManagementService } from "../../services/ModelManagementService";
import { BaseComponent } from "../Base/BaseComponent";

export interface IDeployFailedContentProps {
    deployName: string;
    startTime: string;
}
export interface IDeployFailedContentState {
    deployLog: string | undefined;
}

export class DeployFailedContent extends BaseComponent<
    IDeployFailedContentProps,
    IDeployFailedContentState, {
        modelManagementService: ModelManagementService;
    }>  {
    protected serviceConstructors = {
        modelManagementService: ModelManagementService
    };

    constructor(props: IDeployFailedContentProps) {
        super(props);
        this.state = {
            deployLog: undefined
        };
    }

    public readonly render = (): React.ReactNode => {
        return (
            <>
                <MessageBar messageBarType={MessageBarType.error}>
                    <strong>{this.props.deployName}</strong> failed to deploy. <br /><br />
                    <Link onClick={this.downloadDeployLog}>Download logs</Link> for details <br /><br />
                    Last deployment started on {this.props.startTime}.
                </MessageBar>
            </>
        );
    }

    protected readonly getData = async () => {
        const deployLog = await this.services.modelManagementService.getDeployLogs(this.props.deployName);
        this.setState({ deployLog });
    }

    private readonly downloadDeployLog = async () => {
        if (!this.state.deployLog) {
            return;
        }
        saveAs(new Blob([this.state.deployLog]), "deployLog.txt");
    }
}
