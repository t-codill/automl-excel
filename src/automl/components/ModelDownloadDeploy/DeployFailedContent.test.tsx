import { shallow, ShallowWrapper } from "enzyme";
import { Link } from "office-ui-fabric-react";
import * as React from "react";
import { reactMouseEvent } from "../../__data__/reactMouseEvent";
import * as saveAsModule from "../../common/utils/saveAs";
import { ModelManagementService } from "../../services/ModelManagementService";
import { DeployFailedContent, IDeployFailedContentProps, IDeployFailedContentState } from "./DeployFailedContent";

jest.mock("../../services/ModelManagementService");

let mockGetDeployLogs: jest.SpyInstance<ReturnType<ModelManagementService["getDeployLogs"]>>;

describe("Deploy Failed Content", () => {
    beforeEach(() => {
        mockGetDeployLogs = jest.spyOn(ModelManagementService.prototype, "getDeployLogs");
    });

    describe("if no log", () => {
        let deployLogs: Promise<string | undefined>;
        let tree: ShallowWrapper<IDeployFailedContentProps, IDeployFailedContentState>;

        beforeEach(async () => {
            deployLogs = Promise.resolve(undefined);
            mockGetDeployLogs.mockReturnValue(deployLogs);

            tree = shallow(<DeployFailedContent
                deployName="AAA"
                startTime="test-start-time"
            />);

            await deployLogs;
        });

        it("should render", async () => {
            expect(tree)
                .toMatchSnapshot();
        });

        it("should not trigger download", async () => {
            const link = tree.find(Link)
                .props();
            const saveAsSpy = jest.spyOn(saveAsModule, "saveAs");
            if (link.onClick) {
                link.onClick(reactMouseEvent);
            }
            expect(saveAsSpy)
                .not
                .toBeCalled();
        });
    });

    describe("if has logs", () => {
        let deployLogs: Promise<string | undefined>;
        let tree: ShallowWrapper<IDeployFailedContentProps, IDeployFailedContentState>;

        beforeEach(async () => {
            deployLogs = Promise.resolve("testLog");
            mockGetDeployLogs.mockReturnValue(deployLogs);

            tree = shallow(<DeployFailedContent
                deployName="AAA"
                startTime="test-start-time"
            />);

            await deployLogs;
        });

        it("should render", async () => {
            expect(tree)
                .toMatchSnapshot();
        });

        it("should trigger download", async () => {
            const link = tree.find(Link)
                .props();
            const saveAsSpy = jest.spyOn(saveAsModule, "saveAs");
            if (link.onClick) {
                link.onClick(reactMouseEvent);
            }
            expect(saveAsSpy)
                .toMatchSnapshot();
        });

    });
});
