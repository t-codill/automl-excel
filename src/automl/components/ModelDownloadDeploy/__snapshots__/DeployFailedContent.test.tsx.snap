// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deploy Failed Content if has logs should render 1`] = `
<Fragment>
  <StyledMessageBarBase
    messageBarType={1}
  >
    <strong>
      AAA
    </strong>
     failed to deploy. 
    <br />
    <br />
    <StyledLinkBase
      onClick={[Function]}
    >
      Download logs
    </StyledLinkBase>
     for details 
    <br />
    <br />
    Last deployment started on 
    test-start-time
    .
  </StyledMessageBarBase>
</Fragment>
`;

exports[`Deploy Failed Content if has logs should trigger download 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Blob {},
      "deployLog.txt",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Deploy Failed Content if no log should render 1`] = `
<Fragment>
  <StyledMessageBarBase
    messageBarType={1}
  >
    <strong>
      AAA
    </strong>
     failed to deploy. 
    <br />
    <br />
    <StyledLinkBase
      onClick={[Function]}
    >
      Download logs
    </StyledLinkBase>
     for details 
    <br />
    <br />
    Last deployment started on 
    test-start-time
    .
  </StyledMessageBarBase>
</Fragment>
`;
